### PR TITLE
[mlir][vector] Fix a `target-rank=0` unrolling

### LINKED ISSUE #64269: [mlir] `mlir::vector::TransferWriteOp::getSource() ` crashed with assertion failure "cast<Ty>() argument of incompatible type!"
MLIR built at commit  b4c54b20
Reproduced with:  
`mlir-opt "--convert-vector-to-scf=full-unroll target-rank=0" temp.mlir`

temp.mlir:
``` milr 

  func.func private @func2(%arg0: vector<22xf16>, %arg1: vector<22xf16>, %arg2: vector<22x22x22xf16>) {
    %c1_i64 = arith.constant 1774520363 : i64
    %c1_i32 = arith.constant 1268727148 : i32 
    %c1 = arith.constant 1 : index 
    %c31 = arith.constant 31 : index
    %6 = tensor.empty(%c1, %c31, %c1) : tensor<?x?x?xf16>
    %alloc_17 = memref.alloc(%c1) : memref<?x22xi32>
    %42 = affine.if affine_set<(d0) : ((d0 + 64) ceildiv 64 >= 0, (d0 - (d0 + 64) - d0 ceildiv 64) floordiv 128 == 0)>(%c31) -> i64 {
      %136 = tensor.empty() : tensor<i1>
      %141 = vector.broadcast %c1_i32 : i32 to vector<3xi32>
      vector.transfer_write %141, %alloc_17[%c1, %c1] {permutation_map = affine_map<(d0, d1) -> (d0)>} : vector<3xi32>, memref<?x22xi32>
      affine.yield %c1_i64 : i64
    } else {
      affine.yield %c1_i64 : i64
    }
    %43 = math.log %6 : tensor<?x?x?xf16>
    return
  }
```
 
trace:
```console
Assertion failed: (isa<To>(Val) && "cast<Ty>() argument of incompatible type!"), function cast, file Casting.h, line 566.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: mlir-opt "--convert-vector-to-scf=full-unroll target-rank=0" temp.mlir
 #0 0x00000001024b0548 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/workspace/build/bin/mlir-opt+0x1002f8548)
 #1 0x00000001024ae7e4 llvm::sys::RunSignalHandlers() (/workspace/build/bin/mlir-opt+0x1002f67e4)
 #2 0x00000001024b0bf4 SignalHandler(int) (/workspace/build/bin/mlir-opt+0x1002f8bf4)
 #3 0x00000001a46cd4c4 (/usr/lib/system/libsystem_platform.dylib+0x1803414c4)
 #4 0x00000001a46b5ee0 (/usr/lib/system/libsystem_pthread.dylib+0x180329ee0)
 #5 0x00000001a45f0340 (/usr/lib/system/libsystem_c.dylib+0x180264340)
 #6 0x00000001a45ef754 (/usr/lib/system/libsystem_c.dylib+0x180263754)
 #7 0x00000001068a64dc mlir::vector::TransferWriteOp::getSource() (.cold.1) (/workspace/build/bin/mlir-opt+0x1046ee4dc)
 #8 0x000000010418854c mlir::vector::TransferWriteOp::getSource() (/workspace/build/bin/mlir-opt+0x101fd054c)
 #9 0x00000001045e0498 (anonymous namespace)::lowering_n_d_unrolled::UnrollTransferWriteConversion::matchAndRewrite(mlir::vector::TransferWriteOp, mlir::PatternRewriter&) const (/workspace/build/bin/mlir-opt+0x102428498)
#10 0x0000000106153b74 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>)::$_2::operator()() const (/workspace/build/bin/mlir-opt+0x103f9bb74)
#11 0x0000000106150e70 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) (/workspace/build/bin/mlir-opt+0x103f98e70)
#12 0x0000000104b992b4 (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() (/workspace/build/bin/mlir-opt+0x1029e12b4)
#13 0x0000000104b96ae4 mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) (/workspace/build/bin/mlir-opt+0x1029deae4)
#14 0x00000001045e79c4 (anonymous namespace)::ConvertVectorToSCFPass::runOnOperation() (/workspace/build/bin/mlir-opt+0x10242f9c4)
#15 0x0000000104b39e34 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/workspace/build/bin/mlir-opt+0x102981e34)
#16 0x0000000104b3a524 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/workspace/build/bin/mlir-opt+0x102982524)
#17 0x0000000104b3c3bc mlir::PassManager::run(mlir::Operation*) (/workspace/build/bin/mlir-opt+0x1029843bc)
#18 0x0000000104b34808 performActions(llvm::raw_ostream&, std::__1::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) (/workspace/build/bin/mlir-opt+0x10297c808)
#19 0x0000000104b33d30 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::$_1>(long, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) (/workspace/build/bin/mlir-opt+0x10297bd30)
#20 0x0000000104bb2be0 mlir::splitAndProcessBuffer(std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) (/workspace/build/bin/mlir-opt+0x1029fabe0)
#21 0x0000000104b2e6e4 mlir::MlirOptMain(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/workspace/build/bin/mlir-opt+0x1029766e4)
#22 0x0000000104b2eb54 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/workspace/build/bin/mlir-opt+0x102976b54)
#23 0x00000001021bb58c main (/workspace/build/bin/mlir-opt+0x10000358c)
#24 0x000000010d879088 

```

### PR DESCRIPTION
**Pull request by rikhuijzer**
Fixes https://github.com/llvm/llvm-project/issues/64269.

With this patch, calling `mlir-opt "-convert-vector-to-scf=full-unroll target-rank=0"` on
```mlir
func.func @main(%vec : vector<2xi32>) {
  %alloc = memref.alloc() : memref<4xi32>
  %c0 = arith.constant 0 : index
  vector.transfer_write %vec, %alloc[%c0] : vector<2xi32>, memref<4xi32>
  return
}
```
will result in
```mlir
module {
  func.func @main(%arg0: vector<2xi32>) {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %alloc = memref.alloc() : memref<4xi32>
    %0 = vector.extract %arg0[0] : i32 from vector<2xi32>
    %1 = vector.broadcast %0 : i32 to vector<i32>
    vector.transfer_write %1, %alloc[%c0] : vector<i32>, memref<4xi32>
    %2 = vector.extract %arg0[1] : i32 from vector<2xi32>
    %3 = vector.broadcast %2 : i32 to vector<i32>
    vector.transfer_write %3, %alloc[%c1] : vector<i32>, memref<4xi32>
    return
  }
}
```

I've also tried to proactively find other `target-rank=0` bugs, but couldn't find any. `options.targetRank` is only used 8 times throughout the `mlir` folder, all inside `VectorToSCF.cpp`. None of the other uses look like they could cause a crash. I've also tried

```mlir
func.func @main(%vec : vector<2xi32>) -> vector<2xi32> {
  %alloc = memref.alloc() : memref<4xindex>
  %c0 = arith.constant 0 : index
  %out = vector.transfer_read %alloc[%c0], %c0 : memref<4xindex>, vector<2xi32>
  return %out : vector<2xi32>
}
```
with `"--convert-vector-to-scf=full-unroll target-rank=0"` and that also didn't crash. (Maybe obvious. I have to admit that I'm not very familiar with these ops.)

### CODE REVIEW

File: mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp [hunk 1]
```diff
@@ -1207,23 +1207,30 @@ struct UnrollTransferWriteConversion
...
   LogicalResult matchAndRewrite(TransferWriteOp xferOp,
                                 PatternRewriter &rewriter) const override {
-    if (xferOp.getVectorType().getRank() <= options.targetRank)
+    VectorType inputVectorTy = xferOp.getVectorType();
+
+    if (inputVectorTy.getRank() <= options.targetRank)
       return failure();
+
+    // When target-rank=0, unrolling would cause the vector input argument
+    // into `transfer_write` to become a scalar.
```
[the remark this answers is not in the document]
Comment (rikhuijzer) [reply mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp]: Thank you for your review!

I've implemented this in 769f314 and updated the text for the first comment in this PR.

File: mlir/test/Conversion/VectorToSCF/vector-to-scf.mlir [hunk 2]
```diff
@@ -748,3 +749,16 @@ func.func @cannot_fully_unroll_transfer_write_of_nd_scalable_vector(%vec: vector
...
   return
 }
+
+//  -----
+
+// TARGET-RANK-ZERO-LABEL: func @unroll_transfer_write_target_rank_zero
+func.func @unroll_transfer_write_target_rank_zero(%vec : vector<2xi32>) {
+  // TARGET-RANK-ZERO: %[[EXTRACTED:.*]] = vector.extract
+  // TARGET-RANK-ZERO: %[[BROADCASTED:.*]] = vector.broadcast %[[EXTRACTED]]
+  // TARGET-RANK-ZERO: vector.transfer_write %[[BROADCASTED]]
```
Comment (c-rhodes) [inline mlir/test/Conversion/VectorToSCF/vector-to-scf.mlir:759]: nit: I would find it helpful if the types and both extract/broadcast/transfer_write sequences were CHECK'ed

**Review by c-rhodes — APPROVED** [review]
I've left one minor nit, but otherwise LGTM, thanks for patch!

**Comment by rikhuijzer** [comment]
> I've left one minor nit, but otherwise LGTM, thanks for patch!

I'll fix that and merge once the tests pass. Thanks for the review!

**Pull request merged**